<a href="https://colab.research.google.com/github/SonnieCodes/DS-Unit-1-Sprint-1-Data-Wrangling-and-Storytelling/blob/master/module2-make-features/LS_DS_112_Make_Features_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200> 

# Assignment:

- Replicate the lesson code.

 - This means that if you haven't followed along already, type out the things that we did in class. Forcing your fingers to hit each key will help you internalize the syntax of what we're doing. Make sure you understand each line of code that you're writing, google things that you don't fully understand.
 - [Lambda Learning Method for DS - By Ryan Herr](https://docs.google.com/document/d/1ubOw9B3Hfip27hF2ZFnW3a3z9xAgrUDRReOEo-FHCVs/edit?usp=sharing)
- Convert the `term` column from string to integer.
- Make a column named `loan_status_is_great`. It should contain the integer 1 if `loan_status` is "Current" or "Fully Paid." Else it should contain the integer 0.
- Make `last_pymnt_d_month` and `last_pymnt_d_year` columns.

In [0]:
import pandas as pd

In [3]:
#Import the data file and view the first five rows

df = pd.read_csv('https://raw.githubusercontent.com/ryanleeallred/datasets/master/Ames%20Housing%20Data/train.csv')
df.head()

Id  MSSubClass MSZoning  ...  SaleType  SaleCondition SalePrice
0   1          60       RL  ...        WD         Normal    208500
1   2          20       RL  ...        WD         Normal    181500
2   3          60       RL  ...        WD         Normal    223500
3   4          70       RL  ...        WD        Abnorml    140000
4   5          60       RL  ...        WD         Normal    250000

[5 rows x 81 columns]

In [4]:
#Adjusting display options, and viewing column data types

pd.set_option('display.max_rows', 100)
pd.set_option('display.max_columns', 100)
df.dtypes

Id                 int64
MSSubClass         int64
MSZoning          object
LotFrontage      float64
LotArea            int64
Street            object
Alley             object
LotShape          object
LandContour       object
Utilities         object
LotConfig         object
LandSlope         object
Neighborhood      object
Condition1        object
Condition2        object
BldgType          object
HouseStyle        object
OverallQual        int64
OverallCond        int64
YearBuilt          int64
YearRemodAdd       int64
RoofStyle         object
RoofMatl          object
Exterior1st       object
Exterior2nd       object
MasVnrType        object
MasVnrArea       float64
ExterQual         object
ExterCond         object
Foundation        object
BsmtQual          object
BsmtCond          object
BsmtExposure      object
BsmtFinType1      object
BsmtFinSF1         int64
BsmtFinType2      object
BsmtFinSF2         int64
BsmtUnfSF          int64
TotalBsmtSF        int64
Heating           object


In [5]:
#Viewing first ten items in a column

df['BedroomAbvGr'].head(10)

0    3
1    3
2    3
3    3
4    4
5    1
6    3
7    3
8    2
9    2
Name: BedroomAbvGr, dtype: int64

In [6]:
#Viewing the first ten rows of LotFrontage

df['LotFrontage'].head(10)

0    65.0
1    80.0
2    68.0
3    60.0
4    84.0
5    85.0
6    75.0
7     NaN
8    51.0
9    50.0
Name: LotFrontage, dtype: float64

In [7]:
#All possible values for LotFrontage

df['LotFrontage'].value_counts(dropna=False)

NaN      259
60.0     143
70.0      70
80.0      69
50.0      57
        ... 
106.0      1
38.0       1
138.0      1
140.0      1
137.0      1
Name: LotFrontage, Length: 111, dtype: int64

In [0]:
#Imorting numpy

import numpy as np

In [10]:
#Determining the type of NaN

type(np.NaN)

float

In [12]:
#Creating a smaller data frame

small_df = df[['TotalBsmtSF', '1stFlrSF', '2ndFlrSF', 'SalePrice']]
small_df

TotalBsmtSF  1stFlrSF  2ndFlrSF  SalePrice
0             856       856       854     208500
1            1262      1262         0     181500
2             920       920       866     223500
3             756       961       756     140000
4            1145      1145      1053     250000
...           ...       ...       ...        ...
1455          953       953       694     175000
1456         1542      2073         0     210000
1457         1152      1188      1152     266500
1458         1078      1078         0     142125
1459         1256      1256         0     147500

[1460 rows x 4 columns]

In [13]:
#Creating a total square footage column

small_df['TotalSquareFootage'] = small_df['TotalBsmtSF'] + small_df['1stFlrSF'] + small_df['2ndFlrSF']
small_df

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


TotalBsmtSF  1stFlrSF  2ndFlrSF  SalePrice  TotalSquareFootage
0             856       856       854     208500                2566
1            1262      1262         0     181500                2524
2             920       920       866     223500                2706
3             756       961       756     140000                2473
4            1145      1145      1053     250000                3343
...           ...       ...       ...        ...                 ...
1455          953       953       694     175000                2600
1456         1542      2073         0     210000                3615
1457         1152      1188      1152     266500                3492
1458         1078      1078         0     142125                2156
1459         1256      1256         0     147500                2512

[1460 rows x 5 columns]

In [15]:
#Creating a price per square foot column

small_df['PricePerSqFt'] = small_df['SalePrice'] / small_df['TotalSquareFootage']
small_df.head()

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


TotalBsmtSF  1stFlrSF  2ndFlrSF  SalePrice  TotalSquareFootage  \
0          856       856       854     208500                2566   
1         1262      1262         0     181500                2524   
2          920       920       866     223500                2706   
3          756       961       756     140000                2473   
4         1145      1145      1053     250000                3343   

   PricePerSqFt  
0     81.254871  
1     71.909667  
2     82.594235  
3     56.611403  
4     74.783129

A high price per square foot could indicate homes in good areas, better views, better school districts, etc...   Basically a more desirable house.
A lower price per square foot is just the opposite.  Possibly located in a high crime area, more rural, less attractive neighborhood, etc...

In [16]:
# Working with strings

#Importing a new dataset with a bash command

!wget https://resources.lendingclub.com/LoanStats_2018Q4.csv.zip

--2020-01-14 20:58:30--  https://resources.lendingclub.com/LoanStats_2018Q4.csv.zip
Resolving resources.lendingclub.com (resources.lendingclub.com)... 64.48.1.20
Connecting to resources.lendingclub.com (resources.lendingclub.com)|64.48.1.20|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [application/zip]
Saving to: ‘LoanStats_2018Q4.csv.zip’

LoanStats_2018Q4.cs     [                 <=>]  21.72M   853KB/s    in 26s     

2020-01-14 20:58:57 (843 KB/s) - ‘LoanStats_2018Q4.csv.zip’ saved [22776791]



In [17]:
#Unzipping above file

!unzip LoanStats_2018Q4.csv

Archive:  LoanStats_2018Q4.csv.zip
  inflating: LoanStats_2018Q4.csv    


In [19]:
#Viewing the first 5 rows with a bash command

!head LoanStats_2018Q4.csv

Notes offered by Prospectus (https://www.lendingclub.com/info/prospectus.action)
"id","member_id","loan_amnt","funded_amnt","funded_amnt_inv","term","int_rate","installment","grade","sub_grade","emp_title","emp_length","home_ownership","annual_inc","verification_status","issue_d","loan_status","pymnt_plan","url","desc","purpose","title","zip_code","addr_state","dti","delinq_2yrs","earliest_cr_line","inq_last_6mths","mths_since_last_delinq","mths_since_last_record","open_acc","pub_rec","revol_bal","revol_util","total_acc","initial_list_status","out_prncp","out_prncp_inv","total_pymnt","total_pymnt_inv","total_rec_prncp","total_rec_int","total_rec_late_fee","recoveries","collection_recovery_fee","last_pymnt_d","last_pymnt_amnt","next_pymnt_d","last_credit_pull_d","collections_12_mths_ex_med","mths_since_last_major_derog","policy_code","application_type","annual_inc_joint","dti_joint","verification_status_joint","acc_now_delinq","tot_coll_amt","tot_cur_bal","open_acc_6m","open_act_il","op

In [20]:
#Viewing the last 5 rows with a bash command

!tail LoanStats_2018Q4.csv

"","","5600","5600","5600"," 36 months"," 13.56%","190.21","C","C1","","n/a","RENT","15600","Not Verified","Oct-2018","Current","n","","","credit_card","Credit card refinancing","836xx","ID","15.31","0","Aug-2012","0","","97","9","1","5996","34.5%","11","w","3686.37","3686.37","2652.39","2652.39","1913.63","738.76","0.0","0.0","0.0","Dec-2019","190.21","Jan-2020","Nov-2019","0","","1","Individual","","","","0","0","5996","0","0","0","1","20","0","","0","2","3017","35","17400","1","0","0","3","750","4689","45.5","0","0","20","73","13","13","0","13","","20","","0","3","5","4","4","1","9","10","5","9","0","0","0","0","100","25","1","0","17400","5996","8600","0","","","","","","","","","","","","N","","","","","","","","","","","","","","","N","","","","","",""
"","","23000","23000","23000"," 36 months"," 15.02%","797.53","C","C3","Tax Consultant","10+ years","MORTGAGE","75000","Source Verified","Oct-2018","Charged Off","n","","","debt_consolidation","Debt consolidation","352xx","AL","20.9

Viewing the last five rows shows a potential problem with the data file.  Those last two lines will be read into the dataset unintentionally.

In [21]:
#Reading in the data, printing the shape, and printing the first five rows

df = pd.read_csv('LoanStats_2018Q4.csv')
print(df.shape)
df.head()

/usr/local/lib/python3.6/dist-packages/IPython/core/interactiveshell.py:2718: DtypeWarning: Columns (0,1,2,3,4,7,13,18,19,24,25,27,28,29,30,31,32,34,36,37,38,39,40,41,42,43,44,46,49,50,51,53,54,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,113,114,115,116,117,118,119,120,121,126,127,131,132,134,135,136,141,142,143) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


(128400, 1)


Notes offered by Prospectus (https://www.lendingclub.com/info/prospectus.action)
id  member_id loan_amnt funded_amnt funded_amnt_inv term       int_rate installment grade sub_grade emp_title           emp_length home_ownership annual_inc verification_status issue_d  loan_status pymnt_plan url desc purpose            title                   zip_code addr_state dti   delinq_2yrs earliest_cr_line inq_last_6mths mths_since_last_delinq mths_since_last_record open_acc pub_rec revol_bal revol_util total_acc initial_list_status out_prncp out_prncp_inv total_pymnt     total_pymnt_inv total_rec_prncp total_rec_int total_rec_late_fee recoveries collection_recovery_fee last_pymnt_d last_pymnt_amnt next_pymnt_d last_credit_pull_d collections_12_mths_ex_med mths_since_last_major_derog policy_code application_type annual_inc_joint dti_joint verification_status_joint acc_now_delinq tot_coll_amt tot_cur_bal open_acc_6m open_act_il open_il_12m open_il_24m mths_since_rcnt_il total_bal_il il_util open_rv_12m open_rv_24m max_bal_bc all_util total_rev_hi_lim inq_fi total_cu_tl inq_last_12m acc_open_past_24mths avg_cur_bal bc_open_to_buy bc_util chargeoff_within_12_mths delinq_amnt mo_sin_old_il_acct mo_sin_old_rev_tl_op mo_sin_rcnt_rev_tl_op mo_sin_rcnt_tl mort_acc mths_since_recent_bc mths_since_recent_bc_dlq mths_since_recent_inq mths_since_recent_revol_delinq num_accts_ever_120_pd num_actv_bc_tl num_actv_rev_tl num_bc_sats num_bc_tl num_il_tl num_op_rev_tl num_rev_accts num_rev_tl_bal_gt_0 num_sats num_tl_120dpd_2m num_tl_30dpd num_tl_90g_dpd_24m num_tl_op_past_12m pct_tl_nvr_dlq percent_bc_gt_75 pub_rec_bankruptcies tax_liens tot_hi_cred_lim total_bal_ex_mort total_bc_limit total_il_high_credit_limit revol_bal_joint sec_app_earliest_cr_line sec_app_inq_last_6mths sec_app_mort_acc sec_app_open_acc sec_app_revol_util sec_app_open_act_il sec_app_num_rev_accts sec_app_chargeoff_within_12_mths sec_app_collections_12_mths_ex_med sec_app_mths_since_last_major_derog hardship_flag hardship_type hardship_reason hardship_status deferral_term hardship_amount hardship_start_date hardship_end_date payment_plan_start_date hardship_length hardship_dpd hardship_loan_status orig_projected_additional_accrued_interest hardship_payoff_balance_amount hardship_last_payment_amount debt_settlement_flag debt_settlement_flag_date settlement_status settlement_date settlement_amount settlement_percentage                                    settlement_term                              
NaN NaN       6000      6000        6000             36 months  14.47%  206.44      C     C2        Manager             10+ years  RENT           55000      Verified            Dec-2018 Charged Off n          NaN NaN  credit_card        Credit card refinancing 104xx    NY         10.65 0           Jun-2006         0              NaN                    NaN                    6        0       4196      91.2%      8         w                   0.00      0.00          1106.52         1106.52         546.14          272.38        0.0                288.0      51.84                   Apr-2019     206.44          NaN          Nov-2019           0                          NaN                         1           Individual       NaN              NaN       NaN                       0              0            12514       0           2           1           1           9                  8318         74      1           1           2807       79       4600             1      0           2            2                    2086        344            92      0                        0           33                 150                  9                     9              0        9                    NaN                      9                     NaN                            0                     3              4               3           3         3         4             5             4                   6        0                0            0                  2                  100            100        

The header and last two lines of the data file will need to be fixed

In [22]:
#Fixing the header

df = pd.read_csv('LoanStats_2018Q4.csv', header=1)
df.head()

/usr/local/lib/python3.6/dist-packages/IPython/core/interactiveshell.py:2718: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  NaN        NaN     6000.0       6000.0           6000.0   36 months   
1  NaN        NaN    24000.0      24000.0          24000.0   36 months   
2  NaN        NaN    30000.0      30000.0          30000.0   36 months   
3  NaN        NaN     8000.0       8000.0           8000.0   36 months   
4  NaN        NaN    22400.0      22400.0          22400.0   60 months   

  int_rate  installment grade sub_grade                emp_title emp_length  \
0   14.47%       206.44     C        C2                  Manager  10+ years   
1   10.33%       778.14     B        B1      Maintenance Manager  10+ years   
2   15.02%      1040.26     C        C3             Firefighter   10+ years   
3   12.98%       269.48     B        B5               Controller    4 years   
4   16.14%       546.40     C        C4  Paramedic/ Fire Fighter  10+ years   

  home_ownership  annual_inc verification_status   issue_d  loan_status  \
0           RENT     55000.0            Verified  Dec-2018  Charged Off   
1       MORTGAGE    100000.0        Not Verified  Dec-2018      Current   
2           RENT     90000.0            Verified  Dec-2018   Fully Paid   
3       MORTGAGE     95000.0     Source Verified  Dec-2018      Current   
4            OWN     56000.0     Source Verified  Dec-2018      Current   

  pymnt_plan  url  desc             purpose                    title zip_code  \
0          n  NaN   NaN         credit_card  Credit card refinancing    104xx   
1          n  NaN   NaN  debt_consolidation       Debt consolidation    606xx   
2          n  NaN   NaN  debt_consolidation       Debt consolidation    681xx   
3          n  NaN   NaN         credit_card  Credit card refinancing    480xx   
4          n  NaN   NaN         credit_card  Credit card refinancing    234xx   

  addr_state    dti  delinq_2yrs earliest_cr_line  inq_last_6mths  \
0         NY  10.65          0.0         Jun-2006             0.0   
1         IL  19.14          0.0         Oct-1996             3.0   
2         NE  22.23          0.0         May-2007             1.0   
3         MI   9.44          0.0         Mar-2009             2.0   
4         VA  18.47          0.0         Dec-2004             0.0   

   mths_since_last_delinq  mths_since_last_record  open_acc  pub_rec  \
0                     NaN                     NaN       6.0      0.0   
1                    39.0                     NaN      14.0      0.0   
2                     NaN                     NaN       8.0      0.0   
3                    73.0                     NaN       7.0      0.0   
4                    75.0                     NaN      15.0      0.0   

   revol_bal revol_util  total_acc initial_list_status  out_prncp  \
0     4196.0      91.2%        8.0                   w       0.00   
1     8547.0      27.7%       28.0                   w   16807.30   
2    25754.0        48%       23.0                   w       0.00   
3    11434.0      33.3%       20.0                   w    1707.34   
4    13555.0      38.2%       22.0                   w   19230.91   

   out_prncp_inv   total_pymnt  total_pymnt_inv  total_rec_prncp  \
0           0.00   1106.520000          1106.52           546.14   
1       16807.30   9317.020000          9317.02          7192.70   
2           0.00  30883.673712         30883.67         30000.00   
3        1707.34   6855.630000          6855.63          6292.66   
4       19230.91   6526.670000          6526.67          3169.09   

   total_rec_int  total_rec_late_fee  recoveries  collection_recovery_fee  \
0         272.38                 0.0       288.0                    51.84   
1        2124.32                 0.0         0.0                     0.00   
2         883.67                 0.0         0.0                     0.00   
3         562.97                 0.0         0.0                     0.00   
4        3357.58                 0.0         0.0                     0.00   

  last_pymn

In [24]:
#Looking at the NaN values of each column to see the problem that the extra rows at the bottom of the file create

pd.set_option('display.max_rows', 200)
df.isnull().sum().sort_values()

inq_fi                                             2
delinq_amnt                                        2
chargeoff_within_12_mths                           2
acc_open_past_24mths                               2
inq_last_12m                                       2
total_cu_tl                                        2
total_rev_hi_lim                                   2
max_bal_bc                                         2
open_rv_24m                                        2
open_rv_12m                                        2
total_bal_il                                       2
open_il_24m                                        2
hardship_flag                                      2
open_act_il                                        2
open_acc_6m                                        2
tot_cur_bal                                        2
tot_coll_amt                                       2
acc_now_delinq                                     2
application_type                              

Lots of 2s...

In [25]:
#Addressing the extra NaNs in each column by skipping the footer as well

df = pd.read_csv('LoanStats_2018Q4.csv', header=1, skipfooter=2, engine='python')
print (df.shape)
df.head()

(128397, 144)


id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0 NaN        NaN       6000         6000           6000.0   36 months   
1 NaN        NaN      24000        24000          24000.0   36 months   
2 NaN        NaN      30000        30000          30000.0   36 months   
3 NaN        NaN       8000         8000           8000.0   36 months   
4 NaN        NaN      22400        22400          22400.0   60 months   

  int_rate  installment grade sub_grade                emp_title emp_length  \
0   14.47%       206.44     C        C2                  Manager  10+ years   
1   10.33%       778.14     B        B1      Maintenance Manager  10+ years   
2   15.02%      1040.26     C        C3             Firefighter   10+ years   
3   12.98%       269.48     B        B5               Controller    4 years   
4   16.14%       546.40     C        C4  Paramedic/ Fire Fighter  10+ years   

  home_ownership  annual_inc verification_status   issue_d  loan_status  \
0           RENT     55000.0            Verified  Dec-2018  Charged Off   
1       MORTGAGE    100000.0        Not Verified  Dec-2018      Current   
2           RENT     90000.0            Verified  Dec-2018   Fully Paid   
3       MORTGAGE     95000.0     Source Verified  Dec-2018      Current   
4            OWN     56000.0     Source Verified  Dec-2018      Current   

  pymnt_plan  url  desc             purpose                    title zip_code  \
0          n  NaN   NaN         credit_card  Credit card refinancing    104xx   
1          n  NaN   NaN  debt_consolidation       Debt consolidation    606xx   
2          n  NaN   NaN  debt_consolidation       Debt consolidation    681xx   
3          n  NaN   NaN         credit_card  Credit card refinancing    480xx   
4          n  NaN   NaN         credit_card  Credit card refinancing    234xx   

  addr_state    dti  delinq_2yrs earliest_cr_line  inq_last_6mths  \
0         NY  10.65            0         Jun-2006               0   
1         IL  19.14            0         Oct-1996               3   
2         NE  22.23            0         May-2007               1   
3         MI   9.44            0         Mar-2009               2   
4         VA  18.47            0         Dec-2004               0   

   mths_since_last_delinq  mths_since_last_record  open_acc  pub_rec  \
0                     NaN                     NaN         6        0   
1                    39.0                     NaN        14        0   
2                     NaN                     NaN         8        0   
3                    73.0                     NaN         7        0   
4                    75.0                     NaN        15        0   

   revol_bal revol_util  total_acc initial_list_status  out_prncp  \
0       4196      91.2%          8                   w       0.00   
1       8547      27.7%         28                   w   16807.30   
2      25754        48%         23                   w       0.00   
3      11434      33.3%         20                   w    1707.34   
4      13555      38.2%         22                   w   19230.91   

   out_prncp_inv   total_pymnt  total_pymnt_inv  total_rec_prncp  \
0           0.00   1106.520000          1106.52           546.14   
1       16807.30   9317.020000          9317.02          7192.70   
2           0.00  30883.673712         30883.67         30000.00   
3        1707.34   6855.630000          6855.63          6292.66   
4       19230.91   6526.670000          6526.67          3169.09   

   total_rec_int  total_rec_late_fee  recoveries  collection_recovery_fee  \
0         272.38                 0.0       288.0                    51.84   
1        2124.32                 0.0         0.0                     0.00   
2         883.67                 0.0         0.0                     0.00   
3         562.97                 0.0         0.0                     0.00   
4        3357.58                 0.0         0.0                     0.00   

  last_pymnt_d  

In [26]:
#Check our work

df.isnull().sum().sort_values()

inq_fi                                             0
delinq_amnt                                        0
chargeoff_within_12_mths                           0
acc_open_past_24mths                               0
inq_last_12m                                       0
total_cu_tl                                        0
total_rev_hi_lim                                   0
max_bal_bc                                         0
open_rv_24m                                        0
open_rv_12m                                        0
total_bal_il                                       0
open_il_24m                                        0
hardship_flag                                      0
open_act_il                                        0
open_acc_6m                                        0
tot_cur_bal                                        0
tot_coll_amt                                       0
acc_now_delinq                                     0
application_type                              

The above no longer has 2 NaN columns

---



In [27]:
#dropping columns that are entirely NaN values
#Otherwise dropping all columns whose length matches the total file length of 128397

df = df.drop(['url', 'member_id', 'desc', 'id'], axis=1)
df.head()

loan_amnt  funded_amnt  funded_amnt_inv        term int_rate  installment  \
0       6000         6000           6000.0   36 months   14.47%       206.44   
1      24000        24000          24000.0   36 months   10.33%       778.14   
2      30000        30000          30000.0   36 months   15.02%      1040.26   
3       8000         8000           8000.0   36 months   12.98%       269.48   
4      22400        22400          22400.0   60 months   16.14%       546.40   

  grade sub_grade                emp_title emp_length home_ownership  \
0     C        C2                  Manager  10+ years           RENT   
1     B        B1      Maintenance Manager  10+ years       MORTGAGE   
2     C        C3             Firefighter   10+ years           RENT   
3     B        B5               Controller    4 years       MORTGAGE   
4     C        C4  Paramedic/ Fire Fighter  10+ years            OWN   

   annual_inc verification_status   issue_d  loan_status pymnt_plan  \
0     55000.0            Verified  Dec-2018  Charged Off          n   
1    100000.0        Not Verified  Dec-2018      Current          n   
2     90000.0            Verified  Dec-2018   Fully Paid          n   
3     95000.0     Source Verified  Dec-2018      Current          n   
4     56000.0     Source Verified  Dec-2018      Current          n   

              purpose                    title zip_code addr_state    dti  \
0         credit_card  Credit card refinancing    104xx         NY  10.65   
1  debt_consolidation       Debt consolidation    606xx         IL  19.14   
2  debt_consolidation       Debt consolidation    681xx         NE  22.23   
3         credit_card  Credit card refinancing    480xx         MI   9.44   
4         credit_card  Credit card refinancing    234xx         VA  18.47   

   delinq_2yrs earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0            0         Jun-2006               0                     NaN   
1            0         Oct-1996               3                    39.0   
2            0         May-2007               1                     NaN   
3            0         Mar-2009               2                    73.0   
4            0         Dec-2004               0                    75.0   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN         6        0       4196      91.2%          8   
1                     NaN        14        0       8547      27.7%         28   
2                     NaN         8        0      25754        48%         23   
3                     NaN         7        0      11434      33.3%         20   
4                     NaN        15        0      13555      38.2%         22   

  initial_list_status  out_prncp  out_prncp_inv   total_pymnt  \
0                   w       0.00           0.00   1106.520000   
1                   w   16807.30       16807.30   9317.020000   
2                   w       0.00           0.00  30883.673712   
3                   w    1707.34        1707.34   6855.630000   
4                   w   19230.91       19230.91   6526.670000   

   total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0          1106.52           546.14         272.38                 0.0   
1          9317.02          7192.70        2124.32                 0.0   
2         30883.67         30000.00         883.67                 0.0   
3          6855.63          6292.66         562.97                 0.0   
4          6526.67          3169.09        3357.58                 0.0   

   recoveries  collection_recovery_fee last_pymnt_d  last_pymnt_amnt  \
0       288.0                    51.84     Apr-2019           206.44   
1         0.0                     0.00     Dec-2019           778.14   
2         0.0                     0.00     Feb-2019         28382.51   
3         0.0                     0.00     Dec-2019           269.48   
4         0.0                     0.00     Dec-2019           546

In [28]:
#Cleaning up the int_rate column

df['int_rate'].head(10)

0     14.47%
1     10.33%
2     15.02%
3     12.98%
4     16.14%
5     12.98%
6     15.02%
7     16.14%
8     14.47%
9     13.56%
Name: int_rate, dtype: object

We can see the above column is made up of strings.  We would like floats

In [0]:
#Starting by fixing one thing at a time
#Creating a variable

int_rate = ' 12.98%'

In [30]:
#Using .strip() to remove blank spaces

int_rate.strip()

'12.98%'

In [32]:
#Using .strip() to remove the percentage sign

int_rate.strip('%')

' 12.98'

In [34]:
#Combining the two

int_rate.strip().strip('%')

'12.98'

In [35]:
#Casting a string to a float

float('8.75')

8.75

In [36]:
#Casting our string valie to a float

float(int_rate.strip().strip('%'))

12.98

In [37]:
#Checking data type

type(float(int_rate.strip().strip('%')))

float

In [0]:
#We now know how to successfully convert the oritional strings to floats
#Building a function to convert as above

def int_rate_to_float(cell_contents):
  return float(cell_contents.strip().strip('%'))

In [39]:
#Testing the function

int_rate_to_float(int_rate)

12.98

In [40]:
#Testing the data type

type(int_rate_to_float(int_rate))

float

In [0]:
#Applying the function to every cell in a column

new_list = []
for cell in df['int_rate']:
  new_list.append(int_rate_to_float(cell))

In [42]:
#Viewing the first 10 valies in the new list

new_list[0:10]

[14.47, 10.33, 15.02, 12.98, 16.14, 12.98, 15.02, 16.14, 14.47, 13.56]

In [43]:
#Turning the new list into a series object, and checking the first 5

df['int_rate_cleaned'] = pd.Series(new_list)
df.head()

loan_amnt  funded_amnt  funded_amnt_inv        term int_rate  installment  \
0       6000         6000           6000.0   36 months   14.47%       206.44   
1      24000        24000          24000.0   36 months   10.33%       778.14   
2      30000        30000          30000.0   36 months   15.02%      1040.26   
3       8000         8000           8000.0   36 months   12.98%       269.48   
4      22400        22400          22400.0   60 months   16.14%       546.40   

  grade sub_grade                emp_title emp_length home_ownership  \
0     C        C2                  Manager  10+ years           RENT   
1     B        B1      Maintenance Manager  10+ years       MORTGAGE   
2     C        C3             Firefighter   10+ years           RENT   
3     B        B5               Controller    4 years       MORTGAGE   
4     C        C4  Paramedic/ Fire Fighter  10+ years            OWN   

   annual_inc verification_status   issue_d  loan_status pymnt_plan  \
0     55000.0            Verified  Dec-2018  Charged Off          n   
1    100000.0        Not Verified  Dec-2018      Current          n   
2     90000.0            Verified  Dec-2018   Fully Paid          n   
3     95000.0     Source Verified  Dec-2018      Current          n   
4     56000.0     Source Verified  Dec-2018      Current          n   

              purpose                    title zip_code addr_state    dti  \
0         credit_card  Credit card refinancing    104xx         NY  10.65   
1  debt_consolidation       Debt consolidation    606xx         IL  19.14   
2  debt_consolidation       Debt consolidation    681xx         NE  22.23   
3         credit_card  Credit card refinancing    480xx         MI   9.44   
4         credit_card  Credit card refinancing    234xx         VA  18.47   

   delinq_2yrs earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0            0         Jun-2006               0                     NaN   
1            0         Oct-1996               3                    39.0   
2            0         May-2007               1                     NaN   
3            0         Mar-2009               2                    73.0   
4            0         Dec-2004               0                    75.0   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN         6        0       4196      91.2%          8   
1                     NaN        14        0       8547      27.7%         28   
2                     NaN         8        0      25754        48%         23   
3                     NaN         7        0      11434      33.3%         20   
4                     NaN        15        0      13555      38.2%         22   

  initial_list_status  out_prncp  out_prncp_inv   total_pymnt  \
0                   w       0.00           0.00   1106.520000   
1                   w   16807.30       16807.30   9317.020000   
2                   w       0.00           0.00  30883.673712   
3                   w    1707.34        1707.34   6855.630000   
4                   w   19230.91       19230.91   6526.670000   

   total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0          1106.52           546.14         272.38                 0.0   
1          9317.02          7192.70        2124.32                 0.0   
2         30883.67         30000.00         883.67                 0.0   
3          6855.63          6292.66         562.97                 0.0   
4          6526.67          3169.09        3357.58                 0.0   

   recoveries  collection_recovery_fee last_pymnt_d  last_pymnt_amnt  \
0       288.0                    51.84     Apr-2019           206.44   
1         0.0                     0.00     Dec-2019           778.14   
2         0.0                     0.00     Feb-2019         28382.51   
3         0.0                     0.00     Dec-2019           269.48   
4         0.0                     0.00     Dec-2019           546

In [44]:
#Checking the data type of our new column

df.dtypes[-5:]

settlement_date           object
settlement_amount        float64
settlement_percentage    float64
settlement_term          float64
int_rate_cleaned         float64
dtype: object

In [45]:
#Improving the code using the .apply() function

df['int_rate_cleaned_from_apply'] = df['int_rate'].apply(int_rate_to_float)
df.head()

loan_amnt  funded_amnt  funded_amnt_inv        term int_rate  installment  \
0       6000         6000           6000.0   36 months   14.47%       206.44   
1      24000        24000          24000.0   36 months   10.33%       778.14   
2      30000        30000          30000.0   36 months   15.02%      1040.26   
3       8000         8000           8000.0   36 months   12.98%       269.48   
4      22400        22400          22400.0   60 months   16.14%       546.40   

  grade sub_grade                emp_title emp_length home_ownership  \
0     C        C2                  Manager  10+ years           RENT   
1     B        B1      Maintenance Manager  10+ years       MORTGAGE   
2     C        C3             Firefighter   10+ years           RENT   
3     B        B5               Controller    4 years       MORTGAGE   
4     C        C4  Paramedic/ Fire Fighter  10+ years            OWN   

   annual_inc verification_status   issue_d  loan_status pymnt_plan  \
0     55000.0            Verified  Dec-2018  Charged Off          n   
1    100000.0        Not Verified  Dec-2018      Current          n   
2     90000.0            Verified  Dec-2018   Fully Paid          n   
3     95000.0     Source Verified  Dec-2018      Current          n   
4     56000.0     Source Verified  Dec-2018      Current          n   

              purpose                    title zip_code addr_state    dti  \
0         credit_card  Credit card refinancing    104xx         NY  10.65   
1  debt_consolidation       Debt consolidation    606xx         IL  19.14   
2  debt_consolidation       Debt consolidation    681xx         NE  22.23   
3         credit_card  Credit card refinancing    480xx         MI   9.44   
4         credit_card  Credit card refinancing    234xx         VA  18.47   

   delinq_2yrs earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0            0         Jun-2006               0                     NaN   
1            0         Oct-1996               3                    39.0   
2            0         May-2007               1                     NaN   
3            0         Mar-2009               2                    73.0   
4            0         Dec-2004               0                    75.0   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN         6        0       4196      91.2%          8   
1                     NaN        14        0       8547      27.7%         28   
2                     NaN         8        0      25754        48%         23   
3                     NaN         7        0      11434      33.3%         20   
4                     NaN        15        0      13555      38.2%         22   

  initial_list_status  out_prncp  out_prncp_inv   total_pymnt  \
0                   w       0.00           0.00   1106.520000   
1                   w   16807.30       16807.30   9317.020000   
2                   w       0.00           0.00  30883.673712   
3                   w    1707.34        1707.34   6855.630000   
4                   w   19230.91       19230.91   6526.670000   

   total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0          1106.52           546.14         272.38                 0.0   
1          9317.02          7192.70        2124.32                 0.0   
2         30883.67         30000.00         883.67                 0.0   
3          6855.63          6292.66         562.97                 0.0   
4          6526.67          3169.09        3357.58                 0.0   

   recoveries  collection_recovery_fee last_pymnt_d  last_pymnt_amnt  \
0       288.0                    51.84     Apr-2019           206.44   
1         0.0                     0.00     Dec-2019           778.14   
2         0.0                     0.00     Feb-2019         28382.51   
3         0.0                     0.00     Dec-2019           269.48   
4         0.0                     0.00     Dec-2019           546

Both the long way, and .apply() produce identical results

In [46]:
#Cleaning up the emp_title column
#Looking at the top 20 employment titles

df['emp_title'].value_counts(dropna=False)[:20]

NaN                   20942
Teacher                2090
Manager                1773
Registered Nurse        951
Driver                  924
RN                      726
Supervisor              697
Sales                   580
Project Manager         526
General Manager         523
Office Manager          521
Owner                   420
Director                402
Operations Manager      387
Truck Driver            387
Nurse                   326
Engineer                325
Sales Manager           304
manager                 301
Supervisor              270
Name: emp_title, dtype: int64

In [47]:
#Finding the number of current unique employmet titles

df['emp_title'].value_counts()

Teacher                               2090
Manager                               1773
Registered Nurse                       951
Driver                                 924
RN                                     726
                                      ... 
ALSS Seat Mechanic                       1
Facilities coordinator and trainer       1
Talent Acquisition Director              1
Safety Compliance Manager                1
Network Communications Tech              1
Name: emp_title, Length: 43889, dtype: int64

In [48]:
#Determining how often the employent_title is null

df['emp_title'].isnull().sum()

20942

Employent titles may have been omitted for privacy reasons, perhaps the person is unemployed, maybe they work for themselves, etc...

In [0]:
#Examples of titles we want to clean up

examples = ['owner', 'Supervisor', ' Project Manager', np.NaN]

In [58]:
#Defining a function to clean up titles

def clean_title(job_title):
  if isinstance(job_title, str):
    return job_title.title().strip()
  else:
    return 'Unknown'

for title in examples:
  print(clean_title(title))

Owner
Supervisor
Project Manager
Unknown


In [59]:
#List comprehensions can combine function calls and for loops over lists
#into one succinct and fairly readable single line of code.

[clean_title(title) for title in examples]

['Owner', 'Supervisor', 'Project Manager', 'Unknown']

In [60]:
#We have a function that works as expected.  Lets apply it to our column.
#This tiem we'll overwrite the original column

df['emp_title'] = df['emp_title'].apply(clean_title)
df.head()

loan_amnt  funded_amnt  funded_amnt_inv        term int_rate  installment  \
0       6000         6000           6000.0   36 months   14.47%       206.44   
1      24000        24000          24000.0   36 months   10.33%       778.14   
2      30000        30000          30000.0   36 months   15.02%      1040.26   
3       8000         8000           8000.0   36 months   12.98%       269.48   
4      22400        22400          22400.0   60 months   16.14%       546.40   

  grade sub_grade                emp_title emp_length home_ownership  \
0     C        C2                  Manager  10+ years           RENT   
1     B        B1      Maintenance Manager  10+ years       MORTGAGE   
2     C        C3              Firefighter  10+ years           RENT   
3     B        B5               Controller    4 years       MORTGAGE   
4     C        C4  Paramedic/ Fire Fighter  10+ years            OWN   

   annual_inc verification_status   issue_d  loan_status pymnt_plan  \
0     55000.0            Verified  Dec-2018  Charged Off          n   
1    100000.0        Not Verified  Dec-2018      Current          n   
2     90000.0            Verified  Dec-2018   Fully Paid          n   
3     95000.0     Source Verified  Dec-2018      Current          n   
4     56000.0     Source Verified  Dec-2018      Current          n   

              purpose                    title zip_code addr_state    dti  \
0         credit_card  Credit card refinancing    104xx         NY  10.65   
1  debt_consolidation       Debt consolidation    606xx         IL  19.14   
2  debt_consolidation       Debt consolidation    681xx         NE  22.23   
3         credit_card  Credit card refinancing    480xx         MI   9.44   
4         credit_card  Credit card refinancing    234xx         VA  18.47   

   delinq_2yrs earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0            0         Jun-2006               0                     NaN   
1            0         Oct-1996               3                    39.0   
2            0         May-2007               1                     NaN   
3            0         Mar-2009               2                    73.0   
4            0         Dec-2004               0                    75.0   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN         6        0       4196      91.2%          8   
1                     NaN        14        0       8547      27.7%         28   
2                     NaN         8        0      25754        48%         23   
3                     NaN         7        0      11434      33.3%         20   
4                     NaN        15        0      13555      38.2%         22   

  initial_list_status  out_prncp  out_prncp_inv   total_pymnt  \
0                   w       0.00           0.00   1106.520000   
1                   w   16807.30       16807.30   9317.020000   
2                   w       0.00           0.00  30883.673712   
3                   w    1707.34        1707.34   6855.630000   
4                   w   19230.91       19230.91   6526.670000   

   total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0          1106.52           546.14         272.38                 0.0   
1          9317.02          7192.70        2124.32                 0.0   
2         30883.67         30000.00         883.67                 0.0   
3          6855.63          6292.66         562.97                 0.0   
4          6526.67          3169.09        3357.58                 0.0   

   recoveries  collection_recovery_fee last_pymnt_d  last_pymnt_amnt  \
0       288.0                    51.84     Apr-2019           206.44   
1         0.0                     0.00     Dec-2019           778.14   
2         0.0                     0.00     Feb-2019         28382.51   
3         0.0                     0.00     Dec-2019           269.48   
4         0.0                     0.00     Dec-2019           546

In [61]:
#Viewing our progress with the top 20 employment titles

df['emp_title'].value_counts(ascending=False)[:20]

Unknown               20942
Teacher                2557
Manager                2395
Registered Nurse       1417
Driver                 1258
Supervisor             1160
Truck Driver            920
Rn                      834
Office Manager          804
Sales                   803
General Manager         791
Project Manager         720
Owner                   625
Director                523
Operations Manager      518
Sales Manager           500
Police Officer          440
Nurse                   424
Technician              420
Engineer                412
Name: emp_title, dtype: int64

Note no more NaNs.  They have been converted into 'Unknown'

In [63]:
#How many different unique employment titles are there currently?

len(df['emp_title'].value_counts())

34899

In [66]:
#How often is the employment_title (NaN)?

df['emp_title'].isnull().sum()

0

In [67]:
#Many of the most useful date columns in this dataset have the suffix _d 
#to indicate they correspond to dates.  We'll use a list comprenhension 
#to print them out.

[col for col in df if col.endswith('_d')]

['issue_d', 'last_pymnt_d', 'next_pymnt_d', 'last_credit_pull_d']

In [68]:
#Looking at the string format of the issue_d column

df['issue_d'][:10]

0    Dec-2018
1    Dec-2018
2    Dec-2018
3    Dec-2018
4    Dec-2018
5    Dec-2018
6    Dec-2018
7    Dec-2018
8    Dec-2018
9    Dec-2018
Name: issue_d, dtype: object

In [70]:
#Letting pandas detect this format and translate it into the appropriate datetime object

df['issue_d'] = pd.to_datetime(df['issue_d'], infer_datetime_format=True)
df.dtypes[:15]

loan_amnt                       int64
funded_amnt                     int64
funded_amnt_inv               float64
term                           object
int_rate                       object
installment                   float64
grade                          object
sub_grade                      object
emp_title                      object
emp_length                     object
home_ownership                 object
annual_inc                    float64
verification_status            object
issue_d                datetime64[ns]
loan_status                    object
dtype: object

issue_d is now a datetime object

In [71]:
#Seeing what datetime objects look like

df['issue_d'].iloc[0]

Timestamp('2018-12-01 00:00:00')

In [72]:
df['issue_d'].head().values

array(['2018-12-01T00:00:00.000000000', '2018-12-01T00:00:00.000000000',
       '2018-12-01T00:00:00.000000000', '2018-12-01T00:00:00.000000000',
       '2018-12-01T00:00:00.000000000'], dtype='datetime64[ns]')

In [73]:
#Using the .dt accessor to grab just the year from all cells in issue_d

df['issue_d'].dt.year

0         2018
1         2018
2         2018
3         2018
4         2018
          ... 
128392    2018
128393    2018
128394    2018
128395    2018
128396    2018
Name: issue_d, Length: 128397, dtype: int64

In [74]:
#Now the month
df['issue_d'].dt.month

0         12
1         12
2         12
3         12
4         12
          ..
128392    10
128393    10
128394    10
128395    10
128396    10
Name: issue_d, Length: 128397, dtype: int64

In [76]:
#adding the year and month columns to the datarame

df['issue_year'] = df['issue_d'].dt.year
df['issue_month'] = df['issue_d'].dt.month
df.head()

loan_amnt  funded_amnt  funded_amnt_inv        term int_rate  installment  \
0       6000         6000           6000.0   36 months   14.47%       206.44   
1      24000        24000          24000.0   36 months   10.33%       778.14   
2      30000        30000          30000.0   36 months   15.02%      1040.26   
3       8000         8000           8000.0   36 months   12.98%       269.48   
4      22400        22400          22400.0   60 months   16.14%       546.40   

  grade sub_grade                emp_title emp_length home_ownership  \
0     C        C2                  Manager  10+ years           RENT   
1     B        B1      Maintenance Manager  10+ years       MORTGAGE   
2     C        C3              Firefighter  10+ years           RENT   
3     B        B5               Controller    4 years       MORTGAGE   
4     C        C4  Paramedic/ Fire Fighter  10+ years            OWN   

   annual_inc verification_status    issue_d  loan_status pymnt_plan  \
0     55000.0            Verified 2018-12-01  Charged Off          n   
1    100000.0        Not Verified 2018-12-01      Current          n   
2     90000.0            Verified 2018-12-01   Fully Paid          n   
3     95000.0     Source Verified 2018-12-01      Current          n   
4     56000.0     Source Verified 2018-12-01      Current          n   

              purpose                    title zip_code addr_state    dti  \
0         credit_card  Credit card refinancing    104xx         NY  10.65   
1  debt_consolidation       Debt consolidation    606xx         IL  19.14   
2  debt_consolidation       Debt consolidation    681xx         NE  22.23   
3         credit_card  Credit card refinancing    480xx         MI   9.44   
4         credit_card  Credit card refinancing    234xx         VA  18.47   

   delinq_2yrs earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0            0         Jun-2006               0                     NaN   
1            0         Oct-1996               3                    39.0   
2            0         May-2007               1                     NaN   
3            0         Mar-2009               2                    73.0   
4            0         Dec-2004               0                    75.0   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN         6        0       4196      91.2%          8   
1                     NaN        14        0       8547      27.7%         28   
2                     NaN         8        0      25754        48%         23   
3                     NaN         7        0      11434      33.3%         20   
4                     NaN        15        0      13555      38.2%         22   

  initial_list_status  out_prncp  out_prncp_inv   total_pymnt  \
0                   w       0.00           0.00   1106.520000   
1                   w   16807.30       16807.30   9317.020000   
2                   w       0.00           0.00  30883.673712   
3                   w    1707.34        1707.34   6855.630000   
4                   w   19230.91       19230.91   6526.670000   

   total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0          1106.52           546.14         272.38                 0.0   
1          9317.02          7192.70        2124.32                 0.0   
2         30883.67         30000.00         883.67                 0.0   
3          6855.63          6292.66         562.97                 0.0   
4          6526.67          3169.09        3357.58                 0.0   

   recoveries  collection_recovery_fee last_pymnt_d  last_pymnt_amnt  \
0       288.0                    51.84     Apr-2019           206.44   
1         0.0                     0.00     Dec-2019           778.14   
2         0.0                     0.00     Feb-2019         28382.51   
3         0.0                     0.00     Dec-2019           269.48   
4         0.0                     0.00     Dec-2019        

In [77]:
#Looking at the data types of our new columns

df.dtypes[-2:]

issue_year     int64
issue_month    int64
dtype: object

In [78]:
#Creating a new column - days_from_earliest_credit_to_issue

df['earliest_cr_line'].head()

0    Jun-2006
1    Oct-1996
2    May-2007
3    Mar-2009
4    Dec-2004
Name: earliest_cr_line, dtype: object

In [0]:
df['earliest_cr_line'] = pd.to_datetime(df['earliest_cr_line'], infer_datetime_format=True)

In [80]:
#Ammount of time between the issuance of the loan, to the persons earliest credit line

df['issue_d'] - df['earliest_cr_line']

0         4566 days
1         8096 days
2         4232 days
3         3562 days
4         5113 days
            ...    
128392   12114 days
128393   13241 days
128394    4505 days
128395    3652 days
128396    4413 days
Length: 128397, dtype: timedelta64[ns]

In [81]:
df['days_from_earliest_credit_to_issue'] = (df['issue_d'] - df['earliest_cr_line']).dt.days
df.head()

loan_amnt  funded_amnt  funded_amnt_inv        term int_rate  installment  \
0       6000         6000           6000.0   36 months   14.47%       206.44   
1      24000        24000          24000.0   36 months   10.33%       778.14   
2      30000        30000          30000.0   36 months   15.02%      1040.26   
3       8000         8000           8000.0   36 months   12.98%       269.48   
4      22400        22400          22400.0   60 months   16.14%       546.40   

  grade sub_grade                emp_title emp_length home_ownership  \
0     C        C2                  Manager  10+ years           RENT   
1     B        B1      Maintenance Manager  10+ years       MORTGAGE   
2     C        C3              Firefighter  10+ years           RENT   
3     B        B5               Controller    4 years       MORTGAGE   
4     C        C4  Paramedic/ Fire Fighter  10+ years            OWN   

   annual_inc verification_status    issue_d  loan_status pymnt_plan  \
0     55000.0            Verified 2018-12-01  Charged Off          n   
1    100000.0        Not Verified 2018-12-01      Current          n   
2     90000.0            Verified 2018-12-01   Fully Paid          n   
3     95000.0     Source Verified 2018-12-01      Current          n   
4     56000.0     Source Verified 2018-12-01      Current          n   

              purpose                    title zip_code addr_state    dti  \
0         credit_card  Credit card refinancing    104xx         NY  10.65   
1  debt_consolidation       Debt consolidation    606xx         IL  19.14   
2  debt_consolidation       Debt consolidation    681xx         NE  22.23   
3         credit_card  Credit card refinancing    480xx         MI   9.44   
4         credit_card  Credit card refinancing    234xx         VA  18.47   

   delinq_2yrs earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0            0       2006-06-01               0                     NaN   
1            0       1996-10-01               3                    39.0   
2            0       2007-05-01               1                     NaN   
3            0       2009-03-01               2                    73.0   
4            0       2004-12-01               0                    75.0   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN         6        0       4196      91.2%          8   
1                     NaN        14        0       8547      27.7%         28   
2                     NaN         8        0      25754        48%         23   
3                     NaN         7        0      11434      33.3%         20   
4                     NaN        15        0      13555      38.2%         22   

  initial_list_status  out_prncp  out_prncp_inv   total_pymnt  \
0                   w       0.00           0.00   1106.520000   
1                   w   16807.30       16807.30   9317.020000   
2                   w       0.00           0.00  30883.673712   
3                   w    1707.34        1707.34   6855.630000   
4                   w   19230.91       19230.91   6526.670000   

   total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0          1106.52           546.14         272.38                 0.0   
1          9317.02          7192.70        2124.32                 0.0   
2         30883.67         30000.00         883.67                 0.0   
3          6855.63          6292.66         562.97                 0.0   
4          6526.67          3169.09        3357.58                 0.0   

   recoveries  collection_recovery_fee last_pymnt_d  last_pymnt_amnt  \
0       288.0                    51.84     Apr-2019           206.44   
1         0.0                     0.00     Dec-2019           778.14   
2         0.0                     0.00     Feb-2019         28382.51   
3         0.0                     0.00     Dec-2019           269.48   
4         0.0                     0.00     Dec-2019        

In [82]:
#Finding the oldeest credit history in Q4 2018
df['days_from_earliest_credit_to_issue'].describe()

count    128397.000000
mean       5859.918464
std        2886.617678
min        1126.000000
25%        4049.000000
50%        5266.000000
75%        7244.000000
max       25171.000000
Name: days_from_earliest_credit_to_issue, dtype: float64

25,171 days(68.96 years) is the longest credit history

# Stretch Goals

You can do more with the LendingClub or Instacart datasets.

LendingClub options:
- There's one other column in the dataframe with percent signs. Remove them and convert to floats. You'll need to handle missing values.
- Modify the `emp_title` column to replace titles with 'Other' if the title is not in the top 20. 
- Take initiatve and work on your own ideas!

Instacart options:
- Read [Instacart Market Basket Analysis, Winner's Interview: 2nd place, Kazuki Onodera](http://blog.kaggle.com/2017/09/21/instacart-market-basket-analysis-winners-interview-2nd-place-kazuki-onodera/), especially the **Feature Engineering** section. (Can you choose one feature from his bulleted lists, and try to engineer it with pandas code?)
- Read and replicate parts of [Simple Exploration Notebook - Instacart](https://www.kaggle.com/sudalairajkumar/simple-exploration-notebook-instacart). (It's the Python Notebook with the most upvotes for this Kaggle competition.)
- Take initiative and work on your own ideas!

You can uncomment and run the cells below to re-download and extract the Instacart data

In [86]:
#There's one other column in the dataframe with percent signs. 
#Remove them and convert to floats. You'll need to handle missing values.

df.head()

loan_amnt  funded_amnt  funded_amnt_inv        term int_rate  installment  \
0       6000         6000           6000.0   36 months   14.47%       206.44   
1      24000        24000          24000.0   36 months   10.33%       778.14   
2      30000        30000          30000.0   36 months   15.02%      1040.26   
3       8000         8000           8000.0   36 months   12.98%       269.48   
4      22400        22400          22400.0   60 months   16.14%       546.40   

  grade sub_grade                emp_title emp_length home_ownership  \
0     C        C2                  Manager  10+ years           RENT   
1     B        B1      Maintenance Manager  10+ years       MORTGAGE   
2     C        C3              Firefighter  10+ years           RENT   
3     B        B5               Controller    4 years       MORTGAGE   
4     C        C4  Paramedic/ Fire Fighter  10+ years            OWN   

   annual_inc verification_status    issue_d  loan_status pymnt_plan  \
0     55000.0            Verified 2018-12-01  Charged Off          n   
1    100000.0        Not Verified 2018-12-01      Current          n   
2     90000.0            Verified 2018-12-01   Fully Paid          n   
3     95000.0     Source Verified 2018-12-01      Current          n   
4     56000.0     Source Verified 2018-12-01      Current          n   

              purpose                    title zip_code addr_state    dti  \
0         credit_card  Credit card refinancing    104xx         NY  10.65   
1  debt_consolidation       Debt consolidation    606xx         IL  19.14   
2  debt_consolidation       Debt consolidation    681xx         NE  22.23   
3         credit_card  Credit card refinancing    480xx         MI   9.44   
4         credit_card  Credit card refinancing    234xx         VA  18.47   

   delinq_2yrs earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0            0       2006-06-01               0                     NaN   
1            0       1996-10-01               3                    39.0   
2            0       2007-05-01               1                     NaN   
3            0       2009-03-01               2                    73.0   
4            0       2004-12-01               0                    75.0   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN         6        0       4196      91.2%          8   
1                     NaN        14        0       8547      27.7%         28   
2                     NaN         8        0      25754        48%         23   
3                     NaN         7        0      11434      33.3%         20   
4                     NaN        15        0      13555      38.2%         22   

  initial_list_status  out_prncp  out_prncp_inv   total_pymnt  \
0                   w       0.00           0.00   1106.520000   
1                   w   16807.30       16807.30   9317.020000   
2                   w       0.00           0.00  30883.673712   
3                   w    1707.34        1707.34   6855.630000   
4                   w   19230.91       19230.91   6526.670000   

   total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0          1106.52           546.14         272.38                 0.0   
1          9317.02          7192.70        2124.32                 0.0   
2         30883.67         30000.00         883.67                 0.0   
3          6855.63          6292.66         562.97                 0.0   
4          6526.67          3169.09        3357.58                 0.0   

   recoveries  collection_recovery_fee last_pymnt_d  last_pymnt_amnt  \
0       288.0                    51.84     Apr-2019           206.44   
1         0.0                     0.00     Dec-2019           778.14   
2         0.0                     0.00     Feb-2019         28382.51   
3         0.0                     0.00     Dec-2019           269.48   
4         0.0                     0.00     Dec-2019        

In [99]:
#revol_util has percentage signs

df['revol_util'].head()

0    91.2%
1    27.7%
2      48%
3    33.3%
4    38.2%
Name: revol_util, dtype: object

In [109]:
#Testing a .stip() function to remove % and convert to float

type(float(df['revol_util'][2].strip('%')))

float

In [0]:
#Function for converting as above

def revol_util_to_float(cell_contents):
  return float(cell_contents.strip('%'))

In [110]:
#Testing the function
revol_util_to_float(df['revol_util'][2])

48.0

In [118]:
#Number of null cells

df['revol_util'].isnull().sum()

156

In [0]:
#Replacing NaN with '0'

df['revol_util'].fillna('0', inplace=True)

In [123]:
#Rechecking for null values

df['revol_util'].isnull().sum()

0

In [0]:
#Applying the function to every cell in the column

df['revol_util'] = df['revol_util'].apply(revol_util_to_float)

In [121]:
#Checking for the proper solution

df['revol_util'].head()

0    91.2
1    27.7
2    48.0
3    33.3
4    38.2
Name: revol_util, dtype: float64

In [127]:
#Modify the emp_title column to replace titles with 'Other' if the title is not in the top 20.

#Viewing the top 20 employment titles
#Note--This list views the top 21, however 'Unknown' is at the top of the list,
#so we will view 21 with the intention of ignoring 'Unknown'

df['emp_title'].value_counts(dropna=False)[:21]

Unknown                     20942
Teacher                      2557
Manager                      2395
Registered Nurse             1417
Driver                       1258
Supervisor                   1160
Truck Driver                  920
Rn                            834
Office Manager                804
Sales                         803
General Manager               791
Project Manager               720
Owner                         625
Director                      523
Operations Manager            518
Sales Manager                 500
Police Officer                440
Nurse                         424
Technician                    420
Engineer                      412
Administrative Assistant      409
Name: emp_title, dtype: int64

In [173]:
#Another method of finding the top 20 using Counter

from collections import Counter
c = Counter(df['emp_title'])
c.most_common(21)

[('Unknown', 20942),
 ('Teacher', 2557),
 ('Manager', 2395),
 ('Registered Nurse', 1417),
 ('Driver', 1258),
 ('Supervisor', 1160),
 ('Truck Driver', 920),
 ('Rn', 834),
 ('Office Manager', 804),
 ('Sales', 803),
 ('General Manager', 791),
 ('Project Manager', 720),
 ('Owner', 625),
 ('Director', 523),
 ('Operations Manager', 518),
 ('Sales Manager', 500),
 ('Police Officer', 440),
 ('Nurse', 424),
 ('Technician', 420),
 ('Engineer', 412),
 ('Administrative Assistant', 409)]

In [0]:
!wget https://s3.amazonaws.com/instacart-datasets/instacart_online_grocery_shopping_2017_05_01.tar.gz

In [0]:
!tar --gunzip --extract --verbose --file=instacart_online_grocery_shopping_2017_05_01.tar.gz

In [0]:
%cd instacart_2017_05_01